libhttpserver/server.go

```diff
@@ -0,0 +1,170 @@
+// Copyright 2018 Keybase Inc. All rights reserved.
+// Use of this source code is governed by a BSD
+// license that can be found in the LICENSE file.
+
+package libhttpserver
+
+import (
+	"context"
+	"crypto/rand"
+	"encoding/hex"
+	"errors"
+	"io"
+	"net/http"
+	"path"
+	"strings"
+
+	"github.com/hashicorp/golang-lru"
+	"github.com/keybase/client/go/libkb"
+	"github.com/keybase/client/go/protocol/keybase1"
+	"github.com/keybase/kbfs/libfs"
+	"github.com/keybase/kbfs/libkbfs"
+	"github.com/keybase/kbfs/libmime"
+	"github.com/keybase/kbfs/tlf"
+)
+
+const tokenCacheSize = 64
+const fsCacheSize = 64
+
+// Server is a local HTTP server for serving KBFS content to the front end.
+type Server struct {
+	config libkbfs.Config
+	server *libkb.HTTPSrv
+
+	tokens *lru.Cache
+	fs     *lru.Cache
+}
+
+var _ libkbfs.LocalHTTPServer = (*Server)(nil)
+
+const tokenByteSize = 16
+
+// NewToken implements the libkbfs.LocalHTTPServer interface..
+func (s *Server) NewToken() (token string, err error) {
+	buf := make([]byte, tokenByteSize)
+	if _, err = rand.Read(buf); err != nil {
+		return "", err
+	}
+	token = hex.EncodeToString(buf)
+	s.tokens.Add(token, nil)
+	return token, nil
+}
+
+func (s *Server) handleInvalidToken(w http.ResponseWriter) {
+	w.WriteHeader(http.StatusForbidden)
+	io.WriteString(w, `
+    <html>
+        <head>
+            <title>KBFS HTTP Token Invalid</title>
+        </head>
+        <body>
+            token invalid
+        </body>
+    </html>
+    `)
+}
+
+func (s *Server) handleBadRequest(w http.ResponseWriter) {
+	w.WriteHeader(http.StatusBadRequest)
+}
+
+func (s *Server) getHTTPFileSystem(ctx context.Context, requestPath string) (
+	toStrip string, fs http.FileSystem, err error) {
+	fields := strings.Split(requestPath, "/")
+	if len(fields) < 3 {
+		return "", nil, errors.New("bad path")
+	}
+
+	tlfType, err := tlf.ParseTlfType(fields[0])
+	if err != nil {
+		return "", nil, err
+	}
+
+	tlfHandle, err := libkbfs.GetHandleFromFolderNameAndType(ctx,
+		s.config.KBPKI(), s.config.MDOps(), fields[1], tlfType)
+	if err != nil {
+		return "", nil, err
+	}
+
+	toStrip = path.Join(fields[0], fields[1])
+
+	fav := tlfHandle.ToFavorite()
+	if fsCached, ok := s.fs.Get(fav); ok {
+		if tlfFS, ok := fsCached.(*libfs.FS); ok {
+			return toStrip, tlfFS.ToHTTPFileSystem(ctx), nil
+		}
+	}
+
+	tlfFS, err := libfs.NewFS(ctx,
+		s.config, tlfHandle, "", "", keybase1.MDPriorityNormal)
+	if err != nil {
+		return "", nil, err
+	}
+	s.fs.Add(fav, tlfFS)
+
+	return toStrip, tlfFS.ToHTTPFileSystem(ctx), nil
+}
+
+// serve accetps "/<fs path>?token=<token>"
+// For example:
+//     /team/keybase/file.txt?token=1234567890abcdef1234567890abcdef
+func (s *Server) serve(w http.ResponseWriter, req *http.Request) {
+	token := req.URL.Query().Get("token")
+	if len(token) == 0 || !s.tokens.Contains(token) {
+		s.handleInvalidToken(w)
+		return
+	}
+	toStrip, fs, err := s.getHTTPFileSystem(req.Context(), req.URL.Path)
```

Comment:
Are tokens going to be unique to a session or unique to a file?  If the latter, should we be deleting the token from the cache now that it's used?    If the former, do we have to worry about some other program eavesdropping on the "http" localhost connection and reusing the same tokens?

Reply:
This limits the HTTP server's access to those that can access the service socket, meaning the user running the app and `root`. If we didn't have the token mechanism, other users on the same machine can request content easily.

(I was wrong about enabling HSTS. That's irrelevant actually.)

Reply:
assuming the OS doesn't allow non-root user to snoop traffic, which I think is the default

Reply:
Ah ok.  So other non-root users on the same machine definitely can't listen to regular HTTP traffic on the `lo` interface?

Reply:
I think you need certain permissions to do that. It's possible to grant non-root user that permissions though: https://askubuntu.com/a/632189